streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java

```diff
@@ -49,48 +52,233 @@ public StreamTask restoringTaskFor(final TopicPartition partition) {
+                        log.error(
+                            "After suspending failed, closing the same {} {} failed again due to the following error:",
+                            taskTypeName, id, f);
+                    }
+                } finally {
+                    it.remove();
+
+                    // Wait to reset changelog partitions and remove from running/runningByPartition until the task is
+                    // actually closed as it may be reassigned again, unless an error occurred while suspending
+                    if (!suspended.containsKey(id)) {
+                        running.remove(id);
+                        runningByPartition.keySet().removeAll(task.partitions());
+                        revokedChangelogs.addAll(task.changelogPartitions());
+                    }
+                }
+            }
+        }
+        log.trace("Suspended running {} {}", taskTypeName, suspended.keySet());
+        return firstException.get();
+    }
+
+    private RuntimeException closeNonRunningTasks(final Set<TaskId> revokedNonRunningTasks, final List<TopicPartition> revokedChangelogs) {
+        final Set<TaskId> closedNonRunningTasks = new HashSet<>();
         RuntimeException exception = null;
 
-        log.trace("Closing all restoring stream tasks {}", restoring.keySet());
-        final Iterator<StreamTask> restoringTaskIterator = restoring.values().iterator();
-        while (restoringTaskIterator.hasNext()) {
-            final StreamTask task = restoringTaskIterator.next();
-            log.debug("Closing restoring task {}", task.id());
+        for (final Iterator<TaskId> it = revokedNonRunningTasks.iterator(); it.hasNext(); ) {
+            final TaskId id = it.next();
+            final StreamTask task = created.get(id);
+
             try {
-                task.closeStateManager(true);
+                task.close(false, false);
             } catch (final RuntimeException e) {
-                log.error("Failed to remove restoring task {} due to the following error:", task.id(), e);
+                log.error("Failed to close {}, {}", taskTypeName, task.id(), e);
                 if (exception == null) {
                     exception = e;
                 }
             } finally {
-                restoringTaskIterator.remove();
+                it.remove();
+                created.remove(id);
+                revokedChangelogs.addAll(task.changelogPartitions());
+                closedNonRunningTasks.add(task.id);
             }
         }
+        log.trace("Closed created {} {}", taskTypeName, closedNonRunningTasks);
+        return exception;
+    }
 
-        restoring.clear();
-        restoredPartitions.clear();
-        restoringByPartition.clear();
+    RuntimeException closeRestoringTasks(final Set<TaskId> revokedRestoringTasks, final List<TopicPartition> revokedChangelogs) {
+        log.trace("Closing restoring stream tasks {}", revokedRestoringTasks);
```

Comment:
Why do we have a trace for closing restoring tasks while a trace for closed non-running tasks? What's the difference here?

Reply:
Do you mean what is the difference between restoring and non running? "non running" refers to tasks that haven't been initialized yet & are still in `created`